lib/cache/fs_store.go

```diff
@@ -15,40 +15,105 @@
 package cache
 
 import (
+	"encoding/json"
 	"io/ioutil"
 	"os"
-	"path/filepath"
 	"sync"
+	"time"
 )
 
+type cacheEntry struct {
+	LayerSHA  string
+	Timestamp int64
+}
+
 type fsStore struct {
 	sync.Mutex
 
-	root string
+	fullpath   string
+	sandboxDir string
+	ttlsec     int64
+
+	entries map[string]*cacheEntry
 }
 
-// NewFSStore returns a KVStore backed by the local filesystem. Each key stored
-// will correspond to a file on disk and its value is the contents of that key.
-func NewFSStore(root string) KVStore {
-	return &fsStore{
-		root: root,
+// NewFSStore returns a KVStore backed by the local filesystem.
+// Entries are stored in json format.
+// TODO: enforce capacity.
+func NewFSStore(fullpath string, sandboxDir string, ttlsec int64) (KVStore, error) {
+	s := &fsStore{
+		fullpath:   fullpath,
+		sandboxDir: sandboxDir,
+		ttlsec:     ttlsec,
+		entries:    make(map[string]*cacheEntry),
 	}
-}
 
-func (store *fsStore) Get(key string) (string, error) {
-	path := filepath.Join(store.root, key)
-	contents, err := ioutil.ReadFile(path)
+	contents, err := ioutil.ReadFile(fullpath)
 	if os.IsNotExist(err) {
-		return "", nil
+		return s, nil
 	} else if err != nil {
-		return "", err
+		return nil, err
+	}
+	if err := json.Unmarshal(contents, &s.entries); err != nil {
+		if err := os.Remove(fullpath); err != nil {
+			return nil, err
+		}
+		return s, nil
+	}
+
+	// Remove entries that's older than TTL.
+	for key, entry := range s.entries {
+		if time.Now().Unix()-entry.Timestamp > s.ttlsec {
+			// Cache expired.
+			delete(s.entries, key)
+		}
+	}
+
+	return s, nil
+}
+
+func (s *fsStore) Get(key string) (string, error) {
```

Comment:
Should we be locking in Get and Put for thread safety? It looks like we already have a Mutex in there, is it just not being used right now?